Refactor med_methods_FB_init to allow handling of water tracer fields

billsacks opened the pull request:
### Description of changes

Rather than having med_methods_FB_init accept one of fieldNameList or
STflds, now it accepts an array of med_field_info_type objects.
med_field_info_type is defined in a new module that also includes some
creation methods, making calls to med_methods_FB_init a two-step
process: first create field_info_array, then call med_methods_FB_init
with this field_info_array.

A key new feature is that the creation of a field_info_array from
field_names will assume a single ungridded dimension with size given by
shr_wtracers_get_num_tracers for any field with the water tracer suffix.
This was the main motivation for the refactor here.

Some specific notes about the changes in this commit:
- Behavior change: if there are only scalar or blank fields, we do some
  unnecessary work now (retrieving the mesh that we won't end up ever
  using)
- I think I got the gridToFieldMap right: before it was hard-coded to 2,
  but I think that was only right for a single ungridded dimension; here
  I'm generalizing that for multiple ungridded dimensions

**This depends on the CESM_share changes in https://github.com/ESCOMP/CESM_share/pull/78 and https://github.com/ESCOMP/CESM_share/pull/80**

### Specific notes

Contributors other than yourself, if any:

CMEPS Issues Fixed (include github issue #):

Are changes expected to change answers? (specify if bfb, different at roundoff, more substantial) No - bfb

Any User Interface Changes (namelist or namelist defaults changes)? No

### Testing performed
Please describe the tests along with the target model and machine(s) 
If possible, please also added hashes that were used in the testing

Ran a subset of the CESM prealpha tests with baseline comparisons, using cesm3_0_alpha08a with this CMEPS branch and the share branch from https://github.com/ESCOMP/CESM_share/pull/78:

```
ERS_D_Ld3.ne30pg3_t232.B1850C_LTso.derecho_intel.allactive-defaultio
ERI.ne30pg3_t232.B1850C_LTso.derecho_intel.allactive-defaultio
ERC_D_Ln9.ne30pg3_ne30pg3_mt232.FHISTC_LTso.derecho_intel.cam-outfrq9s
ERS_Ly7.f09_g17_gris4.T1850Gg.derecho_intel
SMS_Lm13.f10_f10_mg37.I1850Clm50SpG.derecho_intel
MULTINOAIS_Ly2.f10_f10_ais8gris4_mg37.I1850Clm50SpRsGag.derecho_intel.cism-change_params
SMS_D_Ld1.ne30pg3_t232.I1850Clm50BgcSpinup.derecho_intel.clm-cplhist
ERI.TL319_t232.G_JRA.derecho_intel.cice-default
SMS_D.TL319_t232.G_JRA_RYF.derecho_intel
SMS_Ld40.TL319_t232.C_JRA.derecho_intel
```

All tests pass and are bit-for-bit with the baselines.

There were NLCOMP failures in some tests; from spot-checking, the NLCOMP failures all seem to be in MOM files. Nothing in this PR or the accompanying share PR could change the namelists, so I assume these are a pre-existing issue:

```
    FAIL ERI.ne30pg3_t232.B1850C_LTso.derecho_intel.allactive-defaultio NLCOMP
    FAIL ERI.TL319_t232.G_JRA.derecho_intel.cice-default NLCOMP
    FAIL ERS_D_Ld3.ne30pg3_t232.B1850C_LTso.derecho_intel.allactive-defaultio NLCOMP
    FAIL SMS_D.TL319_t232.G_JRA_RYF.derecho_intel NLCOMP
    FAIL SMS_Ld40.TL319_t232.C_JRA.derecho_intel NLCOMP
```

jedwards4b commented:
Perhaps if the water isotope code is something for cesm only and that will not be used by UFS or other external CDEPS users its usage should be inside #ifdef CESM blocks?

mvertens commented on mediator/med.F90:
```diff
@@ -1749,19 +1752,25 @@ subroutine DataInitialize(gcomp, rc)
                       STgeom=is_local%wrap%NStateImp(n2), &
-                      STflds=is_local%wrap%NStateImp(n1), &
                       name='FBImp'//trim(compname(n1))//'_'//trim(compname(n2)), rc=rc)
                end if
```
This comment applies to every time med_field_info_array_from_XXX is called in the various routines.

billsacks replied:
I had wondered the same thing. But I think it's okay as is: According to Modern Fortran Explained, in the section on "Allocatable dummy arguments": "If the intent is out and the array is allocated on entry, it becomes deallocated."

billsacks commented:
I have pushed some commits that:

(1) Address requests from @mvertens from our discussion yesterday (I have *not* made changes per @mvertens 's above review - see my response to that review above... happy to discuss that more if you disagree with what I said there).

(2) Remove the new dependencies on share code. I have done this by introducing a small wrapper layer on top of shr_wtracers_mod. For CESM, this wraps shr_wtracers_mod; for UFS (and the standalone build, which uses the ufs directory), this stubs out the used function. In a future PR where I more fully implement the water tracer functionality, I'll build out these wrappers, wrapping/stubbing some additional functions.

@jedwards4b - I have added you as a reviewer, mainly thinking that you could review how I introduced these new wrappers: see wtracers_mod.F90 in the cesm and ufs directories, as well as my changes to buildexe.

Note that, for CESM, this depends on the CESM_share changes in https://github.com/ESCOMP/CESM_share/pull/80.

billsacks commented:
> > This makes sense to me but now you have two files of the same name which must have different contents instead of two files of the same name that must have the same contents. Is that any better?
> 
> Hmmm, you raise a good point, and you got me thinking more about this, but I do think this is better, because having two files with the same contents either requires continual effort to keep them in sync or confusion when they get out of sync and someone wonders if it matters that they differ (as has caused me confusion in similar situations in the past). Here it is hopefully clear that the version here is just a stub and so there's no expectation that it be kept in sync any more than is necessary (i.e., adding stubs for any new APIs that are used).
> 
> That said, I'm happy to discuss this further if you have concerns with this approach.

Oh, and I forgot one other point that seems important at least for this specific case: I realized that, in some instances, I may actually want different logic for UFS than for CESM, because in UFS the tracer module won't be properly initialized, so it could actually help to have a stub version that does nothing and/or returns hard-coded values rather than running through the actual logic.
